private/buf/bufsync/syncer.go

```diff
@@ -56,33 +57,112 @@ func newSyncer(
 	return s, nil
 }
 
+// resolveSyncPoints resolves sync points for all known modules for the specified branch,
+// returning all modules for which sync points were found, along with their sync points.
+//
+// If a SyncPointResolver is not configured, this returns an empty map immediately.
+func (s *syncer) resolveSyncPoints(ctx context.Context, branch string) (map[Module]git.Hash, error) {
+	syncPoints := map[Module]git.Hash{}
+	// If resumption is not enabled, we can bail early.
+	if s.syncPointResolver == nil {
+		return syncPoints, nil
+	}
+	for _, module := range s.modulesToSync {
+		syncPoint, err := s.resolveSyncPoint(ctx, module, branch)
+		if err != nil {
+			return nil, err
+		}
+		if syncPoint != nil {
+			s.logger.Debug(
+				"resolved sync point, will sync after this commit",
+				zap.String("branch", branch),
+				zap.Stringer("module", module),
+				zap.Stringer("syncPoint", syncPoint),
+			)
+			syncPoints[module] = syncPoint
+		} else {
+			s.logger.Debug(
+				"no sync point, syncing from the beginning",
+				zap.String("branch", branch),
+				zap.Stringer("module", module),
+			)
+		}
+	}
+	return syncPoints, nil
+}
+
+// resolveSyncPoint resolves a sync point for a particular module and branch. It assumes
+// that a SyncPointResolver is configured.
+func (s *syncer) resolveSyncPoint(ctx context.Context, module Module, branch string) (git.Hash, error) {
+	syncPoint, err := s.syncPointResolver(ctx, module.RemoteIdentity(), branch)
+	if err != nil {
+		return nil, fmt.Errorf("resolve syncPoint for module %s: %w", module.RemoteIdentity(), err)
+	}
+	if syncPoint != nil {
+		// Validate that the commit pointed to by the sync point exists.
+		if _, err := s.repo.Objects().Commit(syncPoint); err != nil {
+			return nil, s.errorHandler.InvalidSyncPoint(module, branch, syncPoint, err)
+		}
+		return syncPoint, nil
+	}
+	return nil, nil
```

Comment:
Another nit to reduce nesting
```suggestion
	if syncPoint == nil {
		return nil, nil
	}
	// Validate that the commit pointed to by the sync point exists.
	if _, err := s.repo.Objects().Commit(syncPoint); err != nil {
		return nil, s.errorHandler.InvalidSyncPoint(module, branch, syncPoint, err)
	}
	return syncPoint, nil
```